feat:#2-Metagene-визуализация для bsxplorer2: HoloViews + Polars/Plot…

Michael14567 opened the pull request:
Реализован масштабируемый интерфейс визуализации метагена поверх backend bsxplorer2 без жёсткой привязки к структуре (promoter–body–terminator).
Использованы RegionReader и HcAnnotStore: формируем произвольные регионы, дискретизируем на стороне Rust (BsxBatch.discretise), визуализируем на Python.
Сделаны четыре графика: Line, Heat map, Box, Violin — в двух режимах: HoloViews (для ноутбуков) и Polars+Plotly (standalone HTML).
Есть end‑to‑end сценарий на реальных тестовых данных (.bsx + .gff) и набор тестов (юнит/интеграция/е2е).
Документация расширена: обзор архитектуры, примеры использования, “quick reference” по параметрам.
Что сделано

Данные (независимые структуры):
python/src/bsx2/plots/data.py: DiscreteRegionData (хранение x/y по регионам), LinePlotData (валидное представление для линий).
Вычисления:
python/src/bsx2/plots/metagene.py:
Segment + helpers (segments_total_bins, segment_boundaries, segment_ticks).
compute_discrete_regions(reader, contigs, ...) — строит DiscreteRegionData через RegionReader/iter_contigs и Rust‑дискретизацию (BsxBatch.discretise).
compute_from_annot(reader, annot, ...) — то же, но берём Contig из HcAnnotStore.
HoloViews: line_plot, heatmap, box_plot, violin_plot (возвращают hv‑объекты).
HTML (standalone):
python/src/bsx2/plots/polars_html.py:
Полярные преобразования: discrete_to_long_pl, line_df, heatmap_df, dist_df.
Отрисовка: line_html, heatmap_html, box_html, violin_html.
Обёртки “из аннотаций”: *_html_from_annot(reader, annot, ...).
Экспорт API:
python/src/bsx2/plots/__init__.py: общий экспорт всего вышеперечисленного.
Документация:
python/README.md: добавлены разделы с архитектурой, примерами (HoloViews/Plotly), quick‑reference параметров (segments, agg_method, limit, feature_type), сценарии тестирования.
Тесты:
Юнит: python/tests/plots/test_polars_html.py (преобразования Polars) — 3 passed.
Интеграция (Rust дискретизация): python/tests/plots/test_integration_rust.py — 1 passed.
End‑to‑end (iter_contigs + .bsx + .gff): python/tests/plots/test_end2end_iter.py — генерирует HTML и проверяет типы HoloViews (skip, если нет данных).
Пример end‑to‑end:
python/end2end_metagene.py — из .bsx + .gff строит 4 HTML и делает HoloViews smoke.
Как это работает (связка с Rust .bsx)

Источник данных .bsx читается через RegionReader (memory‑map + индекс), запросы регионов идут стримингом (iter_contigs).
Для каждого региона получаем BsxBatch, вызываем Rust‑метод discretise(n_bins, AggMethod), который равномерно разбивает интервал и агрегирует плотности (Mean/Median/Max/Min).
Python‑слой:
Приводит профили к единому виду (DiscreteRegionData), ориентирует минус‑цепь, агрегирует по регионам (Polars).
Рисует HoloViews или генерирует Plotly HTML.
Как запустить (WSL/Debian, Python 3.12)

Установка зависимостей (в venv 3.12): pip install polars pyarrow plotly holoviews matplotlib numpy pandas pytest
Развёртывание Python‑модуля: python -m maturin develop --features nightly -vv (из каталога python)
Запуск тестов:
pytest -q python/tests/plots/test_polars_html.py
pytest -q python/tests/plots/test_integration_rust.py
pytest -q python/tests/plots/test_end2end_iter.py (в наличии bsxplorer2/tests/data/report.bsx и annot.gff)
Выбор библиотек

Визуализация:
HoloViews — для ноутбуков/интерактивной работы.
Plotly — для генерации standalone HTML (удобно встраивать, не требует Python на стороне просмотра).
Таблицы: Polars (+ PyArrow).
Тесты: pytest.
Управление зависимостями: Poetry.
Стиль: Ruff (настройки как в v1), docstrings — NumPy‑стиль.
Что не сделано и планируется

Валидация структур и данных (beartype):
Пока не подключали — планируем добавить точечно на “внешних границах” API (compute_* и *_html_from_annot), с опцией включения через флаг окружения (например, BSX2_STRICT_TYPES=1).
Для NumPy параметров возможно использовать Annotated + beartype.vale для проверки размерности/отсортированности/диапазона.
Отдельный pytest.ini для регистрации маркеров (integration/end2end).
CI‑сборка колёс (Linux/macOS/Windows), публикация артефактов. На Windows wheel‑сборка из путей C:\ может включать абсолютные пути в RECORD (ошибка pip) — мы решаем это сборкой в Linux‑FS и/или корректировками packaging.
Производительность:
Для очень больших наборов — потоковая агрегирующая логика на Python‑слое (инкрементальный расчёт средних по бинам), доп. фильтры в RegionReader (coverage/context) перед дискретизацией.
Примечания

Сегменты задаются произвольно (например, [Segment("up", 25), Segment("body", 50), Segment("down", 25)]), это отражено в интерфейсе и визуализации (xticks по границам).
Для HoloViews обязательно вызвать hv.extension('matplotlib') или 'bokeh' перед работой с .opts.
End‑to‑end скрипт python/end2end_metagene.py демонстрирует полный путь от .bsx + .gff до 4 HTML‑файлов и smoke HoloViews.

Michael14567 committed:
feat:#2-Metagene-визуализация для bsxplorer2: HoloViews + Polars/Plotly, end‑to‑end на .bsx + .gff, тесты

Реализован масштабируемый интерфейс визуализации метагена поверх backend bsxplorer2 без жёсткой привязки к структуре (promoter–body–terminator).
Использованы RegionReader и HcAnnotStore: формируем произвольные регионы, дискретизируем на стороне Rust (BsxBatch.discretise), визуализируем на Python.
Сделаны четыре графика: Line, Heat map, Box, Violin — в двух режимах: HoloViews (для ноутбуков) и Polars+Plotly (standalone HTML).
Есть end‑to‑end сценарий на реальных тестовых данных (.bsx + .gff) и набор тестов (юнит/интеграция/е2е).

README изменил добавил обзор архитектуры, примеры использования, “quick reference” по параметрам. Планирую убрать это из финальной версии но для понимания расписал.

shitohana reviewed (commented):
Давай на будущее code comments на английском писать

shitohana commented on python/build_dev.log:
Этот файл в gitignore

shitohana replied:
и удаляем

shitohana commented on python/src/bsx2/plots/data.py:
```diff
@@ -0,0 +1,73 @@
+        assert positions.ndim == 1 and densities.ndim == 1, "positions/densities must be 1D"
+        assert len(positions) == len(densities), "length mismatch positions/densities"
+        if len(positions) > 1:
+            assert np.all(positions[:-1] <= positions[1:]), "positions must be sorted"
```
Давай сделаем python ошибки по типу ValueError и т.п.

shitohana commented on python/src/bsx2/plots/data.py:
```diff
@@ -0,0 +1,73 @@
+    densities: List[np.ndarray] = field(default_factory=list)   # each: (n_bins,)
+    labels: List[Optional[str]] = field(default_factory=list)
+
+    def insert(self, positions: np.ndarray, densities: np.ndarray, label: Optional[str] = None) -> None:
```
Валидацию в подобных функциях можно как раз делать через beartype. Более того, там можно объявить аннотированный тип, который будет автоматически валидировать то сами данные. В нашем случае, поскольку длина нормализована, мы можем для длины (ну и density) валидировать, что все значения находятся в интервале [0;1]

Там же можно добавить валидацию того, что позиции в списке расположены по возрастанию

Michael14567 replied:
Обьеденил изменения выше в коммит Improve plot data validation and clean repo state

shitohana commented on python/src/bsx2/plots/data.py:
```diff
@@ -0,0 +1,73 @@
+        row_labels = [lbl if lbl is not None else f"region_{i+1}" for i, lbl in enumerate(self.labels)]
+        return mat, row_labels
+
+    def to_line_plot(self, agg_fn: Callable = np.nanmean) -> "LinePlotData":
```
Может уберем этот метод и засунем его в отдельный конструктор в LinePlotData. Чтобы DiscreteRegionData был "базовым" классом без зависимости от имплементаций?

Michael14567 replied:
commit fix:#2-Decouple line plot creation from DiscreteRegionData

shitohana commented on python/src/bsx2/plots/data.py:
```diff
@@ -0,0 +1,73 @@
+
+
+@dataclass
+class DiscreteRegionData:
```
Нужно генерализовать этот класс, чтобы он поддерживал пары positions, densities с разными длинами (т.е. первый регион 100, второй 200 и т.д.). Это нужно потому, что bsxplorer поддерживает дискретизацию не только с помощью разбиения на равное количество интервалов, но и с помощью сегментации. Для каждого региона может быть разное кол-во сегментов.

Тогда, соответственно, придется изменить код преобразования в LinePlot

Michael14567 replied:
commit feat:#2-Support ragged regions in plot data

shitohana commented on python/src/bsx2/plots/data.py:
```diff
@@ -0,0 +1,73 @@
+    y_ticks: List[float] = field(default_factory=list)
+    y_labels: List[str] = field(default_factory=list)
+
+    def to_curve(self, x_shift: float = 0.0, y_shift: float = 0.0):
```
Нужно добавить возвращаемый тип

Michael14567 replied:
commit feat:#2-Annotate curve return type

shitohana commented on python/src/bsx2/plots/metagene.py:
```diff
@@ -0,0 +1,255 @@
+from __future__ import annotations
+from dataclasses import dataclass
+from typing import Callable, List, Optional, Sequence, Tuple
+
+import numpy as np
+import pandas as pd
+
+from bsx2 import io as _io
+from bsx2.plots.data import DiscreteRegionData, LinePlotData
+
+
+@dataclass(frozen=True)
+class Segment:
+    """Сегмент метагена с именем и числом бинов."""
+    name: str
+    n_bins: int
+
+
+def segments_total_bins(segments: Sequence[Segment]) -> int:
+    if not segments:
+        raise ValueError("segments must not be empty")
+    if any(s.n_bins <= 0 for s in segments):
+        raise ValueError("each segment n_bins must be > 0")
+    return sum(s.n_bins for s in segments)
+
+
+def segment_boundaries(segments: Sequence[Segment]) -> List[float]:
+    total = segments_total_bins(segments)
+    cum = 0
+    bounds: List[float] = []
+    for s in segments:
+        cum += s.n_bins
+        bounds.append(cum / total)
+    return bounds
+
+
+def segment_ticks(segments: Sequence[Segment]) -> Tuple[List[float], List[str]]:
+    bounds = segment_boundaries(segments)
+    labels = [s.name for s in segments]
+    return bounds, labels
```
Не совсем понимаю, зачем нам нужен этот класс

Michael14567 replied:
Segment — это компактный описатель частей метагена.
Он хранит имя сегмента (name) и желаемое число бинов (n_bins). Он скорее отвечает на вопрос как разбивать регион . Применяется во всех  функциях метагена compute_discrete_regions, compute_from_annot, line_plot, heatmap, box_plot, violin_plot . суммируется общее число бинов (segments_total_bins) для дискретизации batch.discretise(...); 

Почему предлагаю использовать класс , а не кортеж  сейчас явные поля с типами труднее перепутать по порядку, проще читать код и валидировать (например, проверка n_bins > 0 в одном месте). dataclass(frozen=True) даёт удобный __repr__, сравнение и гарантирует неизменяемость входных параметров.
Семантика  позволяет задавать разные сегменты с разным разрешением (например, up/body/down с 25/50/25 бинами или произвольные интервалы при сегментации). Имя сегмента — это подпись на графике; n_bins  его вклад в общую сетку дискретизации. Ну и на будущее если понадобится добавить цвет, вес или тип сегмента, логичнее расширять один структурный объект, чем разносить параллельные списки.


 Формально  можем обойтись и без класса , передавать, например, список кортежей или словарей с {"name": "...", "n_bins": ...} или просто числа и отдельные подписи. Но мне вариант не очень нравится так как придётся менять сигнатуры функций, валидацию , расширять (добавлять поля) будет труднее .

shitohana commented on python/src/bsx2/plots/metagene.py:
```diff
@@ -0,0 +1,255 @@
+                x = 1.0 - x[::-1]
+                y = y[::-1]
+            label = labels[idx] if labels and idx < len(labels) else None
+            data.insert(x, y, label)
```
у reader всегда будет метод iter_contigs и нам приоритетно использовать его, поскольку он читает параллельно и быстрее

shitohana replied:
короче else ветку выпилить нужно

Michael14567 replied:
commit feat:#2-Enforce iter_contigs path for metagene computation

shitohana commented on python/src/bsx2/plots/metagene.py:
```diff
@@ -0,0 +1,255 @@
+
+
+def compute_discrete_regions(
+    reader: _io.RegionReader,
+    contigs: Sequence,
+    *,
+    segments: Sequence[Segment] | None = None,
+    agg_method=None,
+    reverse_negative: bool = True,
+    labels: Optional[Sequence[str]] = None,
```
У нас сейчас дефолтные аргументы получаются спрятаны от внешнего пользователя. Нужно либо сделать их явными дефолтными, либо написать в документации. Я за первый вариант

Michael14567 replied:
commit feat:#2-Expose default segments in metagene API

Michael14567 replied:
Ещё у меня есть предложение  явный default для agg_method
Плюсы - сразу видно из сигнатуры, что по умолчанию используем, например, AggMethod.Mean; меньше сюрпризов для пользователей и автодополнений
Минусы - AggMethod живёт в bsx2._bsx2 (native модуль). Если импортировать его на уровне модуля ради дефолта, это добавит тяжёлый импорт при загрузке bsx2.plots.metagene. В окружениях без собранного расширения это приведёт к ImportError уже при импорте модуля, а не при вызове.


Вариант решения - Оставить agg_method=None, но явно описать в docstring/типах, что по умолчанию → AggMethod.Mean (как сейчас делается внутри функции).

shitohana replied:
если использовать нативный енум проблемно, можно создать обычный питоновский `class AggMethod(enum.Enum)` с теми же вариантами. При обращении к расту будет конвертироваться в растовский тип.

в моем понимании `agg_method = None` - значит мы не аггрегируем

P.S. native модуль, по идее, при сборке в wheel будет включен в пакет, и ImportError не будет, но надо проверить.

shitohana replied:
Кстати, вспомнил, на такой случай есть https://typing.python.org/en/latest/spec/literal.html

Michael14567 committed:
feat#2:error handling

- Replace assert-based validation with explicit exceptions:
  - Type/shape/value checks now raise TypeError / ValueError.
  - Tick/label consistency now raises ValueError.
- Narrow exception handling where appropriate:
  - Attribute access and calls now catch only (AttributeError, TypeError, ValueError).
- Improve missing-dependency reporting:
  - Lazy holoviews import now raises an informative ImportError in plotting helpers.

Files:
- python/src/bsx2/plots/data.py
  - insert: input validation
  - stack_matrix: consistent n_bins check
  - agg_fn validation and aggregation result checks
  - LinePlotData.__post_init__
  - lazy holoviews import with ImportError
- python/src/bsx2/plots/metagene.py
  - _is_negative_strand with narrow exception handling
  - compute_discrete_regions with narrow exception handling
  - heatmap, box_plot, violin_plot now fail with ImportError if holoviews is missing
- python/end2end_metagene.py
  - narrow handling around filter_coverage_gt
  - narrow handling around collect_contigs_from_hcannot
  - strict HoloViews-type checks raising TypeError

Michael14567 committed:
feat#2: Add beartype-based runtime validation for plotting data

Add beartype as an optional runtime type checker:

Updated python/pyproject.toml to include beartype = "^0.18" in [tool.poetry.dependencies].

Introduce beartype-validated array types:

Define Pos1D and Density1D as Annotated[np.ndarray, ...] with custom predicates:

Pos1D: 1D numeric, values in [0, 1], monotonically non-decreasing.

Density1D: 1D numeric, values in [0, 1]; NaN values are allowed.

Apply @beartype to core data structures and helpers:

Cover DiscreteRegionData, LinePlotData and key methods with @beartype.

Simplify manual input checks where they are now enforced by beartype.

Keep cross-parameter consistency checks (e.g. equal lengths of arrays, tick/label alignment) in __post_init__, as these cannot be expressed in type annotations.

Implement soft fallback when beartype is missing:

Import beartype, Annotated, Is with a graceful fallback: if beartype is not installed, the decorator becomes a no-op and annotated types do not break imports.

Recommend installing beartype to enable actual runtime validation.

Touched code:

python/pyproject.toml

Add beartype to Poetry dependencies.

python/src/bsx2/plots/data.py

Import beartype, Annotated, Is with a safe fallback.

Define predicates _is_1d_sorted_unit_positions, _is_1d_unit_density.

Define type aliases Pos1D, Density1D.

Add @beartype to:

DiscreteRegionData

insert(self, positions: Pos1D, densities: Density1D, ...)

stack_matrix

to_line_plot

LinePlotData (fields x: Pos1D, y: Density1D)

to_curve (numeric parameters as float | int)

Keep __post_init__ checks for equal lengths and tick/label consistency.

Notes:

Density1D currently permits NaN in density values; this can be tightened later by adjusting the predicate.

Runtime validation via beartype is active only when the package is installed; otherwise behavior is unchanged aside from the annotations.

shitohana commented:
А можешь сюда еще картинки прикрепить, которые генерятся?

Michael14567 committed:
Improve plot data validation and clean repo state

Enforce 1D, sorted, finite [0..1] arrays via beartype validators in DiscreteRegionData
Keep explicit length checks and raise ValueError on validation failures
Remove ignored build_dev.log from the Python tree

Michael14567 committed:
fix:#2-Decouple line plot creation from DiscreteRegionData

Remove DiscreteRegionData.to_line_plot and add LinePlotData.from_discrete for aggregation
Update metagene line_plot to use the new LinePlotData constructor
Keep DiscreteRegionData focused on data storage and validation

Michael14567 committed:
feat:#2-Support ragged regions in plot data

Allow DiscreteRegionData to handle regions with differing bin counts via a common interpolated grid
Update stack_matrix and LinePlotData.from_discrete to aggregate on the shared grid
Preserve validation and downstream plotting behavior

Michael14567 committed:
feat:#2-Annotate curve return type

 Add explicit return type to LinePlotData.to_curve to document HoloViews output

Michael14567 committed:
feat:#2-Enforce iter_contigs path for metagene computation

Require reader.iter_contigs and drop query fallback in compute_discrete_regions
Keep strand reversal/label handling; skip batches that fail discretization

Michael14567 committed:
feat:#2-Expose default segments in metagene API

Define a shared _DEFAULT_SEGMENTS (single 100-bin region) and use it as the default for segments
Apply the explicit default across compute_discrete_regions, compute_from_annot, and plot helpers

Michael14567 commented:
Картинки 
box html 
<img width="1771" height="822" alt="изображение" src="https://github.com/user-attachments/assets/1c480aa9-7a95-4d09-9e0d-96ec6ffb40f1" />

Michael14567 commented:
Heatmap

<img width="1832" height="772" alt="изображение" src="https://github.com/user-attachments/assets/cbe191c9-3769-4dac-955a-75f69dae087b" />

Michael14567 commented:
line
<img width="1787" height="869" alt="изображение" src="https://github.com/user-attachments/assets/7b99bcc2-2176-4f3f-b31c-d6c4b1df184e" />

Michael14567 commented:
violin 
<img width="1732" height="829" alt="изображение" src="https://github.com/user-attachments/assets/e5ac7202-7b67-4965-97cb-b6ddbb21e48a" />

Michael14567 commented:
По моему все исправлено , по предложению ответь пожалуйста и по классу и готов к  resolve

Michael14567 committed:
feat:#4-Add clustering/PCA computation and Plotly visuals

- Added cluster.py with matrix prep, k-means, PCA, linkage, utilities, and metagene_for_cluster
- Added cluster_vis.py for Plotly HTML (PCA scatter, k-means centroids heatmap, ordered heatmap, dendrogram)
- Exported new APIs via bsx2.plots; tests cover computations and visuals
- Switched imports from src.bsx2 to bsx2 (io.py, types.py, tests) to avoid ModuleNotFoundError outside the source tree; matches installed module in editable/wheel
- Set python/Cargo.toml license to MIT to avoid wheel installation path issues with license-file

Michael14567 committed:
Silence beartype PEP585 warnings in plots.data

Michael14567 committed:
feat#3-Add chromosome methylation map prep and Holoviews visuals

Add chrmap data prep (line/box) with smoothing and per-chromosome ticks/borders
Add Holoviews helpers for line and box/violin plots; generate sample HTML assets from test data
Export new APIs via bsx2.plots and cover with smoke tests

shitohana commented:
Точки на бохплотах убираем по дефолту, опционально можно включить. На последнем бохплоте усы уходят в <0 значения. Этого быть не может, надо обрезать. На лайнплоте не понятно, что нарисовано, но что бы это ни было, разрывов быть не должно, у нас данные последовательны. Хитмап, если там действительно нарисована хромосома, как следует из подписи выглядит неплохо. Только почему там одна строка всего?

shitohana commented on python/src/bsx2/plots/data.py:
```diff
@@ -0,0 +1,166 @@
+from typing import Any, Callable, List, Optional, Tuple
+import numpy as np
+
+# beartype: runtime type checking with value constraints
+try:
+    from beartype import beartype  # type: ignore
+    from beartype.vale import Is  # type: ignore
+    try:  # Python >=3.9 has typing.Annotated; fall back to typing_extensions
+        from typing import Annotated  # type: ignore
+    except Exception:  # pragma: no cover - platform variance
+        from typing_extensions import Annotated  # type: ignore
+except ModuleNotFoundError:  # graceful fallback if beartype is not installed
+    try:
+        from typing import Annotated  # type: ignore
+    except Exception:  # pragma: no cover
+        from typing_extensions import Annotated  # type: ignore
+
+    def beartype(obj):  # type: ignore
+        return obj
+
+    class Is:  # type: ignore
+        def __class_getitem__(cls, item):
+            # Return the predicate itself as metadata for Annotated; no-op without beartype
+            return item
```
Может без этого? В pyproject зафиксируем, что >=3.11 и все

shitohana commented on python/src/bsx2/plots/data.py:
```diff
@@ -0,0 +1,166 @@
+                interp = np.full_like(grid, float(dens[0]))
+            else:
+                # np.interp uses endpoints for out-of-range; positions are already sorted and within [0,1]
+                interp = np.interp(grid, pos, dens)
```
Давай без интерполяции. Это скрытая манипуляция данными, которая может отразиться на дальнейшем анализе, т.е. не отражать реальных данных.

Этот метод необходим для heatmap-а, а для box/violin-plot где он еще используется, как будто не нужен, потому что для box-plot нас интересует общее аггрегированое значение по всему региону, нам не нужны конкретные точки, поэтому там можно было бы просто взять аггрегацию от всех точек региона. 

В хитмапе мы явно знаем "разрешение" к которому мы приводим наши регионы и в таком случае давай действовать так:
Любой регион - это набор точек и координат, мы не ждем, что координаты будут расположены на равном расстоянии друг от друга, поэтому производим простой binning, т.е. при делении на M регионов (для хитмап) в регион m_i попадают точки, чья координата m_(i-1) < x < m_i

shitohana commented on python/src/bsx2/plots/data.py:
```diff
@@ -0,0 +1,166 @@
+
+    @beartype
+    def stack_matrix(self) -> Tuple[np.ndarray, List[str]]:
+        mat, row_labels, _ = self._stack_to_common_grid()
```
А почему нам stack_to_common_grid не перенести полностью в stack_matrix? Или мы планируем stack_matrix как-то усложнять потом?

shitohana commented on python/src/bsx2/plots/polars_html.py:
А нам сильно нужно превращение в Polars?
HoloViews имеет наливную поддержку Plotly. 
Ну или если хочется напрямую Plotly использовать, то все те же графики можно собрать без Polars.

Просто как-будто получается отдельный от metagene модуль, который решает ту же задачу, но по-другому. _(Или я что-то не понял)_

Michael14567 committed:
Use Savitzky-Golay smoothing for chrmap lines

replace moving-average smoothing in prepare_chr_line_data with scipy.signal.savgol_filter

Michael14567 committed:
Merge pull request #6 from shitohana/chr-methylation-map

feat#3-Add chromosome methylation map prep and Holoviews visuals

Michael14567 committed:
Merge pull request #5 from shitohana/clustering-pca-gene-level

Add clustering/PCA pipeline with Plotly visuals and cluster-aware metagene helper

Michael14567 committed:
step

Michael14567 committed:
feat-#2-add nan policy

Added deterministic clustering and NaN policy across PCA/KMeans/linkage; labels stay aligned after filtering/sorting.
Refined demo demo_aw25.py: per‑call RegionReader, contigs sorted via BatchIndex, --seed parameter, chrmap fallback uses BsxFileReader with chrom/size limits, added --max-chroms.
Fixed Plotly/HoloViews options (removed unsupported tools on curves) and length checks in PCA scatter to avoid mismatched arrays.
Chrmap autowindows capped at 1kb, only first N chromosomes, safer aggregation; artifacts now generated end-to-end.

Michael14567 committed:
feat-#2-line_plot

Added weighted methylation support for heatmap/line (value_mode="weighted" computes 100 * sum(mC) / sum(total) per bin).
Added per‑gene NaN filtering via max_nan_frac (drop genes with too many empty bins).
Extended heatmap to support two modes: genes (per‑feature matrix) and aggregate (single-row, BSX1‑style).
Propagated new options (value_mode, max_nan_frac, heatmap_mode) through heatmap_html, heatmap_html_from_annot, and line_html.
Stored per‑point weights (coverage) in DiscreteRegionData and used them in binning.

Commit:
Removed artifacts

shitohana commented on python/rust/io/region.rs:
```diff
@@ -214,14 +214,30 @@ impl PyRegionReaderIterator {
             self.current_index += 1;
 
+            // Debug logging to help track crashes with filters/iter_contigs
+            eprintln!(
```
это можно написать (и нужно) с помощью log::error!

И здесь это скорее warn, а не error
